### DISCUSSION TITLE
Python like relational API in C headers

### DISCUSSION THREAD

**Discussion by ananis25** [duckdb/duckdb#3538] [Ideas]
The python relational API is really ergonomic compared to writing plain sql or pandas expressions. Is it possible to add it to the provided C API? Since C doesn't have classes, one way to expose it would be one function for each relational operator, that takes the reference to a relation object and the expression string to return a reference to the new relation object.

```
rel2 = *filter(*rel, 'i > 1');
rel3 = *project(*rel2, 'i + 1');
```

I write some low level code in the programming language D, which has really convenient interop with C. By wrapping the relational API, it would be really easy to create a native pandas like library for most languages with C FFI. Looking at the code, it seems DuckDbPyRelation is written in C++, would it be possible to similarly extend it to C?

(Apologies if my understanding is incorrect. I've always been uncomfortable with C++ since almost failing the first programming course.)

**Comment by Mytherin** [comment]
It should be possible to add this to the C API. The main thing I'm worried about is memory management. The way this works in the C++ API is that these operators return shared pointers that can be used and freed:

```cpp
shared_ptr<Relation> tbl, filter;

tbl = con.Table("integers");
filter = tbl->Filter("i = 2");
auto result = filter->Execute();
```

The reason these are shared is because you can re-use parts of the query plan, e.g. if you want to do a self-join you can do that as follows:

```cpp
tbl = con.Table("integers");
auto join = tbl->Join(tbl->Alias("t2"), "tbl.i=t2.i");
auto result = join->Execute();
```

How would the memory management for this look like in C? My initial thought is that we use shared pointers internally, and that in C you have to call 'free' on every node you allocate, e.g.

```c
duckdb_database db;
duckdb_connection con;
duckdb_result result;

duckdb_open(":memory:", &db);
duckdb_connect(db, & con);

duckdb_relation tbl, filter;

tbl = duckdb_table(con, "integers");
filter = duckdb_filter(tbl, "i = 2");
result = duckdb_execute(filter);

duckdb_destroy_relation(&filter);
duckdb_destroy_relation(&tbl);

duckdb_destroy_result(&result);
duckdb_disconnect(&connection);
duckdb_close(&database);
```

The way this would work internally is that when you call `duckdb_table`, it returns a pointer to a shared object (with a single outstanding reference). When `tbl` is used in the filter function, `tbl` now has two references, and `filter` is created with one reference. When `duckdb_destroy_relation(&filter)` is called, the `filter` node is destroyed (no outstanding references) and `tbl` has its reference count reduced by one. When `duckdb_destroy_relation(&tbl)` is called, `tbl` is destroyed.

I think this will work, but will not actually be nice to use from within C directly. It's main purpose will be to facilitate other APIs to be created on top of this layer.

An alternative would be to store the relations in the connection, and only destroy them when the connection is destroyed. Perhaps we could have a method to clear all relations from the connection (e.g. `duckdb_clear_relations(connection)`) which destroys all stored relations and makes using them after that point UB. That might be more ergonomic from a user-perspective, but would likely be worse from an API perspective, since you could no longer destroy individual relations.

**Comment by ananis25** [comment]
Thank you for the detailed reply. I hadn't realized relations and tables are equivalent, the "everything is a relation" model makes composition very convenient. 

I like the idea of an explicit API to destroy relations one by one. While inconvenient from C, in most languages the bindings authors can probably create a wrapper class/object around the duckdb relation, and manage its lifetime by calling `duckdb_destroy_relation` in the destructor.
